### PR TITLE
Enhance: Bring back user to 1st page when filters are changed

### PR DESCRIPTION
**Pull request by imanish003**
**URL page number removal on filter change**: When filters are changed, the page number in the URL is removed. This is to ensure that when a filter is applied, the user is taken back to the first page of results. The code accommodates three different formats of page numbers in the URL and removes them accordingly.

Part of #9876 

### Testing

1. Create a new post.
2. Add all filter blocks and a 'Products Beta' block to it.
3. Save the post and navigate to the frontend.
4. Navigate to the 2nd page using pagination. 
5. Modify any filter and confirm that the page number resets to the 1st page upon filter alteration. 
6. Repeat the above steps and validate that this also holds true for:
    - 'Product Collection' block (The product collection block is currently in the experimental phase and is not shipped with the plugin/core)
    - 'Classic WooCommerce Product Grid Block'
    - Note: This behavior doesn't affect the 'All Products' block.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [X] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Enhancement: Bring back user to 1st page when filters are changed

### CODE REVIEW

**Review by kmanijak — APPROVED** [review]
LGTM!

**Review by Aljullu — COMMENTED** [review]
I noticed that when removing all filters, the URL becomes `/?` first and then `/`, causing an unnecessary reload:

[Enregistrament de pantalla des de 2023-06-22 14-56-46.webm](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/bc3a44fa-7474-48a7-9ae3-df22a830a2b0)

Can you reproduce as well?

**Comment by imanish003** [comment]
@Aljullu, I can reproduce the issue. Thank you for identifying this edge case. I will work on fixing it and push the necessary changes 🤝

**Comment by imanish003** [comment]
Hi @Aljullu, I have pushed a fix in [872c583](https://github.com/woocommerce/woocommerce-blocks/pull/9952/commits/872c583b3d868ee6795566820db87faa00e8d404). Can you please take a look again? 🙂

File: assets/js/utils/filters.ts [hunk 1]
```diff
@@ -50,7 +50,19 @@ export function changeUrl( newUrl: string ) {
...
+		/**
+		 * If there are no query parameters in the URL and the URL ends with '?', we remove
+		 * the trailing '?' from the URL. The reason for this is that a '?' in the URL
+		 * typically indicates that query parameters will follow. If there are no query parameters,
+		 * the '?' is unnecessary and can cause the page to reload, which can negatively affect
+		 * performance. Therefore, by removing the '?', we prevent this unnecessary reload.
+		 */
+		const urlObject = new URL( newUrl );
+		const numberOfQueryParams = [ ...urlObject.searchParams.keys() ].length;
+		if ( numberOfQueryParams === 0 && newUrl.endsWith( '?' ) ) {
```
Comment (Aljullu) [inline assets/js/utils/filters.ts]: I think it's safe to skip the query params check (`numberOfQueryParams === 0`), no? I mean, if the URL ends with `?`, it's always safe to remove the last question mark, no matter if there are query params or not, no? :thinking:

**Review by Aljullu — APPROVED** [review]
This is working great, @imanish003! I just left one small comment about a possible small optimization, but besides that, LGTM.

File: assets/js/utils/filters.ts (hunk 1 above)
Comment (imanish003) [reply assets/js/utils/filters.ts]: I think so. I was just trying to be extra careful to ensure I do not break any existing functionality. 

But good point, the trailing "?" can indeed be removed safely without checking for query parameters. I will simplify the code accordingly. Thanks for the suggestion! 🙌🏻

Comment (imanish003) [reply assets/js/utils/filters.ts]: @Aljullu I have made the changes in [9a82cd9](https://github.com/woocommerce/woocommerce-blocks/pull/9952/commits/9a82cd95a27a1031d9e7b78475b334f812e1cf8b). Can you please take a look again just to be sure that everything still works as expected? 🙂

**Review by Aljullu — COMMENTED** [review]
Thanks for the update, @imanish003. I just found one last minor issue: in some occassions I end up with two `&&` in the URL.

Steps to reproduce:

1. Add the Filter by Price and Filter by Attribute blocks alongside the Product Collection block.
2. In the frontend, add a price filter so pagination is still available. The URL will look something like `/?min_price=15`.
3. Using pagination, switch to page two. The URL will look something like `/?query-0-page=2&min_price=15`.
4. Now, add an attribute filter. The URL will look something like `/?min_price=15&&filter_color=blue&query_type_color=or`. Notice the double `&&` before `filter_color`.

Can you reproduce as well?

**Comment by imanish003** [comment]
Hi @Aljullu, This issue isn't reproducing for me 🤔  Here is a quick demo of what I tried:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/698003b8-46da-4cbe-8d23-cfb542a14fd6

If you are still able to reproduce the issue then maybe I can add a logic like this to clean up query params:
```
let url = new URL(newUrl);
let params = new URLSearchParams(url.search);

// Assign the cleaned up search parameters back to the URL
url.search = params.toString();

// Now update newUrl with the cleaned up URL
newUrl = url.toString();
```
What do you think? 🙂

**Review by Aljullu — APPROVED** [review]
Ups, I don't seem to be able to reproduce anymore, sorry for the confusion. :confused: Not sure what was wrong with my setup, I will open an issue if I can reproduce again, but for now, let's merge this PR. :rocket:

**Pull request merged**